frontend/catalyst/jax_tracer.py

```diff
@@ -405,25 +412,123 @@ def trace_observables(obs, qubit_states, p, num_wires, qreg):
+        op_args = jax.numpy.ones(len(obs))
+        jax_obs = trace_hamiltonian(op_args, *nested_obs)
+    elif isinstance(obs, qml.ops.op_math.Prod):
+        nested_obs = [trace_observables(o, qubit_states, p, num_wires, qreg)[0] for o in obs]
+        jax_obs = jprim.tensorobs(*nested_obs)
+    elif isinstance(obs, qml.ops.op_math.SProd):
+        terms = obs.terms()
+        coeffs = jax.numpy.array(terms[0])
+        nested_obs = trace_observables(terms[1][0], qubit_states, p, num_wires, qreg)[0]
+        jax_obs = jprim.hamiltonian(coeffs, nested_obs)
+    else:  # pragma: no cover
         raise RuntimeError(f"unknown observable in measurement process: {obs}")
     return jax_obs, qubits
 
 
+def named_obs(obs, op_args, qubit_states, qreg):
+    """Trace a named observable.
+
+    Args:
+        obs: a named observable
+        op_args: arguments of the observable
+        qubit_states: the statically known qubit state at this program point
+        qreg: the quantum register with the state at this program point
+
+    Returns:
+        jax_obs: a NamedObs JAX primitive used for tracing
+        qubits: a list of qubits used by the observable
+    """
+    assert isinstance(obs, KNOWN_NAMED_OBS)
+    _, wires = op_args
+    qubits = get_qubits_from_wires(wires, qubit_states, qreg)
+    jax_obs = jprim.namedobs(type(obs).__name__, qubits[0])
+    return jax_obs, qubits
+
+
+def hermitian_obs(obs, op_args, qubit_states, qreg):
+    """Trace a Hermitian observable.
+
+    Args:
+        obs: a Hermitian observable
+        op_args: arguments of the observable
+        qubit_states: the statically known qubit state at this program point
+        qreg: the quantum register with the state at this program point
+
+    Returns:
+        jax_obs: a Hermitian JAX primitive used for tracing
+        qubits: a list of qubits used by the observable
+    """
+    assert isinstance(obs, qml.Hermitian)
+    matrix, wires = op_args
+    qubits = get_qubits_from_wires(wires, qubit_states, qreg)
+    jax_obs = jprim.hermitian(matrix, *qubits)
+    return jax_obs, qubits
+
+
+def pauli_sentence_to_rt_obs(paulis, qubit_states, qreg):
+    """Convert a :class:`pennylane.pauli.PauliSentence` into a Hamiltonian.
+
+    Args:
+        paulis: a :class:`pennylane.pauli.PauliSentence`
+        qubit_states: the statically known qubit state at this program point
+        qreg: the quantum register with the state at this program point
+
+    Returns:
+        a Hamiltonian JAX primitive used for tracing
+    """
+    pwords, coeffs = zip(*paulis.items())
+    nested_obs = [pauli_word_to_rt_obs(pword, qubit_states, qreg) for pword in pwords]
+
+    # No need to create a Hamiltonian for a single TensorObs
+    if len(nested_obs) == 1 and coeffs[0] == 1.0:
+        return nested_obs[0]
+
+    coeffs = jax.numpy.asarray(coeffs)
+    return jprim.hamiltonian(coeffs, *nested_obs)
+
+
+def pauli_word_to_rt_obs(obs, qubit_states, qreg):
```

Comment:
```suggestion
def pauli_word_to_tensor(obs, qubit_states, qreg):
```

Reply:
Sure, I had to change the name in some other places. That's the reason I couldn't accept this one. All fixed [8ce5458](https://github.com/PennyLaneAI/catalyst/pull/250/commits/8ce54582d0f8958e7d95979cb3c5fbbbb272aff8)